### PR TITLE
Bugfix: Abilities check final move type instead of default move type

### PR DESCRIPTION
**Pull request by td76099**
## What are the changes?
These changes make it so that `Move` object is passed to abilities instead of `PokemonMove` as well as simplifying some of the logic using for abilities that change move type.

## Why am I doing these changes?
Right now, abilities are searching and pulling Moves every single time because they are (mostly) being passed `PokemonMove` instead of `Move`. They usually call `move.getMove()` over and over again instead of just passing it along. This means that if a move has a type change associated with it because of how the move works or an ability, then other abilities do not take those changes into account because they are just looking up the default type. [This bug report](https://discord.com/channels/1125469663833370665/1244862160862052352) lists it pretty well. There's numerous reports in the `bug-report` channel of this exact issue. #1390 is an instance of this bug as well.

[Asked a lead dev](https://discord.com/channels/1125469663833370665/1176874654015684739/1245735588720541827) before I started testing and this was an approved implementation. A previous implementation was trying to implement it on a need-to-know basis using a new [getType() function](https://github.com/td76099/pokerogue/commit/58bca3687d32f7055688fd8fb007e50abb38cbd0#diff-62e9cf38a6ba2a2b69a102788cdfef2a33cdc173f79124df114636685a51e762R187), but it was recommended to only run those move/attribute applications once [based on this convo](https://discord.com/channels/1125469663833370665/1244059545328947300/1244638318734475335). That led to updating it to make sure they only get ran once by passing the `Move` object instead of looking up the move every time with `move.getMove()`.

[These moves](https://bulbapedia.bulbagarden.net/wiki/Category:Moves_that_change_type) and [these abilities](https://bulbapedia.bulbagarden.net/wiki/Category:Abilities_that_can_modify_move_types) will all properly interact with all of [these defending abilities](https://bulbapedia.bulbagarden.net/wiki/Category:Abilities_that_alter_damage_taken). There will be no more bug reports about Liquid Voice Torch Song not being absorbed by Water Absorb or being improperly affected by Heatproof and more.

## What did change?
`applyMoveAttrs`, `applyPreAttackAbAttrs`, `applyPostAttackAbAttrs`, `applyPreDefendAbAttrs`, and `applyPostDefendAbAttrs` now take in `Move` instead of `PokemonMove` so they don't have to call `move.getMove()` anymore. Therefore, all subsequent classes that extend those base classes have also been updated to use `Move` instead of `PokemonMove`. These will be listed below in the Screenshot/Videos sections where there are tests for every immediate subclass impacted.

These changes also resulted in some small variable changes in `MoveEffectPhase` in `phases.ts` (no logic changes) by using `move` instead of `this.move.getMove()` or `move.type` instead of `this.move.getMove().type`. Likewise in `pokemon.ts` where instead of using `this.move` or `type` it now has `move` and `move.type` respectively. Only "logic" change in `pokemon.ts` is the removal of `VariableMoveTypeAbAttr` because those abilities (Normalize, Pixelate, Galvanize, etc.) have been simplified to use another existing class.

## How to test the changes?
You can pull this PR and test some of the changes listed in the screenshot/video section below using any combo of move type changing abilities and abilities that modify incoming damage based on move type.

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

## Screenshots/Videos
Will be some duplicates and a decent amount of unnecessary sanity checks but better safe than sorry
#### applyMoveAttrs
In `pokemon.ts`:
 - VariableMoveCategoryAttr - none care about move type atm but was impacted by change of `Move` instead of `PokemonMove` in `apply()`
   - [x] Photon Geyser with higher attack

https://github.com/pagefaultgames/pokerogue/assets/85713900/025709c5-de77-4b1c-9619-ad5f424ccc5c

   - [x] Photon Geyser with higher special attack

https://github.com/pagefaultgames/pokerogue/assets/85713900/df3a3fed-76a8-4710-bc8d-98b627a63aa8

 - VariableMoveTypeAttr
   - [x] Water Ivy Cudgel vs Storm Drain and no ability

https://github.com/pagefaultgames/pokerogue/assets/85713900/27cfd598-2b3e-4e64-a9c3-e9f713b1d194

https://github.com/pagefaultgames/pokerogue/assets/85713900/3f39abc6-41c6-414f-923e-57348aac1468

   - [x] Fire Ivy Cudgel vs Thick Fat and no ability

https://github.com/pagefaultgames/pokerogue/assets/85713900/02ff9e23-b92f-4477-8acb-d633bb03f828

https://github.com/pagefaultgames/pokerogue/assets/85713900/55089919-4f16-4ed6-af8d-2c79f6d55f51

   - [x] Rock Ivy Cudgel SE vs Solid Rock and no ability

https://github.com/pagefaultgames/pokerogue/assets/85713900/1bb290a1-2c79-4d8f-be5c-05d9e467e32e

https://github.com/pagefaultgames/pokerogue/assets/85713900/c1d4369b-9ac7-48be-b3cd-34b02017315c

   - [x] Sunny Weather Ball vs Flash Fire and no ability

https://github.com/pagefaultgames/pokerogue/assets/85713900/1f8b4c3c-3eb2-42d9-b792-0dd8d26494db

https://github.com/pagefaultgames/pokerogue/assets/85713900/e7c38559-d92b-429c-8488-da7a95c1cbef

   - [x] Electric Terrain Pulse vs Volt Absorb and no ability

https://github.com/pagefaultgames/pokerogue/assets/85713900/3c43061f-97d0-4db1-96f1-36f21e115594

 - VariableMoveTypeMultiplierAttr - none care about move type atm but was impacted by change of `Move` instead of `PokemonMove` in `apply()`
   - [x] Freeze Dry vs Normal type
   - [x] Freeze Dry vs Water type

https://github.com/pagefaultgames/pokerogue/assets/85713900/552db235-9a7c-4df0-bc0c-dada0cbcc3ca

 - IgnoreWeatherTypeDebuffAttr - none care about move type atm but was impacted by change of `Move` instead of `PokemonMove` in `apply()`
   - [x] Hydro Steam in normal weather
   - [x] Hydro Steam in the sun

https://github.com/pagefaultgames/pokerogue/assets/85713900/cd324f51-9957-423e-838e-8aa7c6b642d2

 - VariablePowerAttr - none care about move type atm but was impacted by change of `Move` instead of `PokemonMove` in `apply()`
   - [x] Grass Knot vs light target
   - [x] Grass Knot vs heavy target

https://github.com/pagefaultgames/pokerogue/assets/85713900/5a84b2cc-132f-4118-bf39-6bc56455c8cc

In `phases.ts`:
 - OverrideMoveEffectAttr - affected by code change from `this.move.getMove()` to just `move`
   - [x] Nature Power based on biome vs no ability
   - [x] Nature Power based on terrain vs no ability
   - [x] Nature Power based on biome vs Volt Absorb
   - [x] Nature Power based on terrain vs Volt Absorb

https://github.com/pagefaultgames/pokerogue/assets/85713900/6e2007de-c66e-4c8e-944e-b3ffac64d51b

 - MultiHitAttr - affected by code change from `this.move.getMove()` to just `move`
   - [x] Bullet Seed vs Sap Sipper
   - [x] Normalize Bullet Seed vs Sap Sipper

https://github.com/pagefaultgames/pokerogue/assets/85713900/072fde07-7cdd-4165-82dc-88ede52ba464

   - [x] Galvanize Fury Attack vs Volt Absorb
   - [x] Galvanize Fury Attack vs Solid Rock (electric)

https://github.com/pagefaultgames/pokerogue/assets/85713900/e1ed609b-f9ac-4afb-970c-799f29f2b5d1

 - MissEffectAttr - affected by code change from `this.move.getMove()` to just `move`
   - [x] Galvanize Thrash vs Volt Absorb - Does not trigger frenzy effect but neither does Normal Thrash vs Ghost

https://github.com/pagefaultgames/pokerogue/assets/85713900/e990f936-c67b-4be7-9f53-e7ba0f439373

   - [x] Galvanize Thrash vs Solid Rock (electric)

https://github.com/pagefaultgames/pokerogue/assets/85713900/3fe20445-6c98-4147-b039-5918ee8782b3

   - [x] High Jump Kick vs normal type/ghost type

https://github.com/pagefaultgames/pokerogue/assets/85713900/176d3595-543d-40e3-b117-48bc158dd879

https://github.com/pagefaultgames/pokerogue/assets/85713900/433032e1-53dc-4877-aca7-8ce8d80ca9d5

   - [x] Normalize High Jump Kick vs Ghost

https://github.com/pagefaultgames/pokerogue/assets/85713900/3fd4236c-91d9-4298-9f68-3801710b3e68

   - [x] Normalize High Jump Kick vs non-Ghost

https://github.com/pagefaultgames/pokerogue/assets/85713900/06cfb936-d6f4-44d5-ae53-53ea44058460

 - NoEffectAttr - affected by code change from `this.move.getMove()` to just `move`
   - [x] High Jump Kick vs normal type/ghost type

https://github.com/pagefaultgames/pokerogue/assets/85713900/176d3595-543d-40e3-b117-48bc158dd879

https://github.com/pagefaultgames/pokerogue/assets/85713900/433032e1-53dc-4877-aca7-8ce8d80ca9d5

   - [x] Normalize High Jump Kick vs Ghost

https://github.com/pagefaultgames/pokerogue/assets/85713900/3fd4236c-91d9-4298-9f68-3801710b3e68

   - [x] Normalize High Jump Kick vs non-Ghost

https://github.com/pagefaultgames/pokerogue/assets/85713900/06cfb936-d6f4-44d5-ae53-53ea44058460

#### applyFilteredMoveAttrs
In `phases.ts`:
 - MoveEffectAttr - affected by code change from `this.move.getMove()` to just `move`
   - Recoil
     - [x] Recoil still works
   - Draining
     - [x] Draining still works
   - Status Effects
     - [x] % chance to burn/paralyze/freeze/poison/etc. still works
   - Stat boosts/reductions
     - [x] % chance to increase/decrease a stat still works

https://github.com/pagefaultgames/pokerogue/assets/85713900/3df19325-5a18-4c13-b608-57a7453b6293

#### applyPreAttackAbAttrs
In `pokemon.ts`:
 - MoveTypeChangeAttr - a couple duplicate sanity checks
   - [x] Normalize vs Ghost type

https://github.com/pagefaultgames/pokerogue/assets/85713900/0e300c25-4290-4a49-aebb-82614976eb19

   - [x] Normalize Ground move vs Flying

https://github.com/pagefaultgames/pokerogue/assets/85713900/c1490b96-d20d-4a49-a6fb-ec09b90ea7a0

   - [x] Normalize Ground move vs Levitate

https://github.com/pagefaultgames/pokerogue/assets/85713900/e1b6f150-1313-4b8b-9003-44b47563ccb5

   - [x] Normalize Ground move vs Earth Eater

https://github.com/pagefaultgames/pokerogue/assets/85713900/565b4227-dd11-4484-b41f-f4158ac58ced

   - [x] Refrigerate vs Thick Fat

https://github.com/pagefaultgames/pokerogue/assets/85713900/8aa45d57-b227-4e63-93f1-7ddc74e89953

   - [x] Galvanize vs Ground type

https://github.com/pagefaultgames/pokerogue/assets/85713900/fd94e802-e1b0-4270-b5df-7df63548ceb8

   - [x] Galvanize vs Volt Absorb

https://github.com/pagefaultgames/pokerogue/assets/85713900/4688dc15-9a2d-4131-ba5f-af21b52c1b16

   - [x] Liquid Voice vs Storm Drain

https://github.com/pagefaultgames/pokerogue/assets/85713900/5fe2d451-0360-46b7-950b-1ab189508faa

   - [x] Sunny Weather Ball vs Thick Fat

https://github.com/pagefaultgames/pokerogue/assets/85713900/baa2d801-d0d3-49be-bcc3-a1f1a6bf4eb3

   - [x] Sunny Weather Ball vs Well Baked Body

https://github.com/pagefaultgames/pokerogue/assets/85713900/6ae44fe7-bf6c-4cc9-836a-17044d17e0ce

 - VariableMovePowerAbAttr - none care about type but was impacted by change from `PokemonMove` to `Move`
   - [x] Technician on a move with less than 60 power
   - [x] Technician on a move with more than 60 power

https://github.com/pagefaultgames/pokerogue/assets/85713900/958a2c77-37fa-46df-96c2-c792c550d9c3

   - [x] Iron Fist with a punching move
   - [x] Iron First with a non-punching move

https://github.com/pagefaultgames/pokerogue/assets/85713900/becbca17-a795-48cd-a538-8c640d616391

 - FieldVariableMovePowerAbAttr - THESE DO NOT WORK IN THE OFFICIAL BRANCH RIGHT NOW
   - [x] Fairy Aura using a Fairy Move vs normal type
   - [x] Fairy Aura using a non-fairy move vs normal type
   - [x] Fairy Aura Misty Terrain Weather Ball vs normal type 
   - [x] non-Fairy Aura Misty Terrain Weather Ball vs normal type
 - DamageBoostAbAttr
   - [x] Tinted Lens Sunny Weather Ball vs Dragon type

https://github.com/pagefaultgames/pokerogue/assets/85713900/f3d6ce55-309c-405e-a8d5-0edd1f986b91

#### applyPreDefendAbAttrs
In `pokemon.ts`:
 - TypeImmunityAbAttr - may overlap with some of the above test cases for sanity
   - [x] Levitate vs Ground Move
   - [x] Levitate vs Ground Move that has been Normalized

https://github.com/pagefaultgames/pokerogue/assets/85713900/65137e31-920b-4056-a0b9-39d0dada2fa0

 - MoveImmunityAbAttr - not dependent on type but was impacted by `PokemonMove` to `Move` change
   - [x] Soundproof vs sound move
   - [x] Soundproof vs non-sound move

https://github.com/pagefaultgames/pokerogue/assets/85713900/1ab15b8b-dea3-4508-8e97-c1a4fb72b5df

 - ReceivedMoveDamageMultiplierAbAttr
   - [x] Fluffy vs Sunny Weather Ball
   - [x] Fluffy vs Contact Move

https://github.com/pagefaultgames/pokerogue/assets/85713900/17c55b4d-b942-4205-903e-415c8aace0a1

   - [x] Solid Rock vs SE Weather Ball
   - [x] Solid Rock vs SE Galvanize

https://github.com/pagefaultgames/pokerogue/assets/85713900/10ebc234-ab1e-491f-92a9-7b55f291ad93

 - FieldPriorityMoveImmunityAbAttr - not dependent on type but was impacted by `PokemonMove` to `Move` change
   - [x] Dazzling vs Priority move
   - [x] Dazzling vs non-priority move

https://github.com/pagefaultgames/pokerogue/assets/85713900/eec53149-170b-40a3-ba82-18fb9a27d4e4

 - PreDefendFullHpEndureAbAttr - not dependent on type but was impacted by `PokemonMove` to `Move` change
   - [x] Sturdy vs overkill move
   - [x] Sturdy vs non-overkill move

https://github.com/pagefaultgames/pokerogue/assets/85713900/cb3d8e64-eeb3-452f-92ee-fe825c3ec8c7

#### applyPostDefendAbAttrs:
In `phases.ts`:
 - PostDefendAbAttr - not dependent on type but was impacted by `PokemonMove` to `Move` change
   - [x] Justified vs Dark type move
   - [x] Justified vs non-Dark type move
   - [x] Justified vs Aura Wheel as Dark and Electric move

https://github.com/pagefaultgames/pokerogue/assets/85713900/d4734b5a-fb09-4cd6-95eb-61bf78d5f2e2

#### applyPostAttackAbAttrs:
In `phases.ts`:
 - PostAttackAbAttr
   - [x] Magician still steals items

https://github.com/pagefaultgames/pokerogue/assets/85713900/b9771456-4953-4035-9df6-528cf5286c0f

### CODE REVIEW

**Comment by td76099** [comment]
@Madmadness65 @molker was hoping to get confirmation that a rework like this for passing the move around (instead of move id to look up the move every time) would be approved so I can start testing a bunch of stuff! Fixes a bug that happens more often than it should.

**Comment by td76099** [comment]
Testing complete! Ready for review.

**Comment by dirtydom236** [comment]
Hello, just did a bit of testing for this pr and I noticed that after using a refrigerated tackle once on the opponent, it straight up turns tackle into an ice move universally. I made sure to use Amaura so that there was no refrigerate override active, and I can confirm that this does not happen on main (third screenshot). 
image
image
image


https://github.com/pagefaultgames/pokerogue/assets/130926160/90b516e1-ba29-4316-89f2-d1e8bbbc3d16

**Comment by td76099** [comment]
Really good catch - I'll take a look at getting that fixed!

**Comment by td76099** [comment]
Alright I have pushed a fix for that issue - thanks for finding that issue! If you test some more and encounter anything else I'll take ASAP @dirtydom236 

https://github.com/pagefaultgames/pokerogue/assets/85713900/6a2d4e98-000c-43d8-9742-0854f100cbfb

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -1654,44 +1650,44 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
-        defendingSidePlayField.forEach((p) => applyPreDefendAbAttrs(FieldPriorityMoveImmunityAbAttr, p, source, battlerMove, cancelled, typeMultiplier));
+        applyPreDefendAbAttrs(MoveImmunityAbAttr, this, source, move, cancelled, typeMultiplier);
+        defendingSidePlayField.forEach((p) => applyPreDefendAbAttrs(FieldPriorityMoveImmunityAbAttr, p, source, move, cancelled, typeMultiplier));
       }
 
       if (cancelled.value) {
         result = HitResult.NO_EFFECT;
       } else {
-        const typeBoost = source.findTag(t => t instanceof TypeBoostTag && (t as TypeBoostTag).boostedType === type) as TypeBoostTag;
+        const typeBoost = source.findTag(t => t instanceof TypeBoostTag && (t as TypeBoostTag).boostedType === move.type) as TypeBoostTag;
```
Comment (OrangeRed) [inline src/field/pokemon.ts:1674]: Can you remove this type assertion here? `t instanceof TypeBoostTag && t.boostedType === move.type` 
Typescript is smart enough to know that t has type `TypeBoostTag` after you do instanceof

File: src/phases.ts [hunk 2]
```diff
@@ -2743,31 +2744,31 @@ export class MoveEffectPhase extends PokemonPhase {
...
                 // Charge attribute with charge effect takes all effect attributes and applies them to charge stage, so ignore them if this is present
                 Utils.executeIf(!chargeEffect, () => applyFilteredMoveAttrs((attr: MoveAttr) => attr instanceof MoveEffectAttr && (attr as MoveEffectAttr).trigger === MoveEffectTrigger.POST_APPLY
-                  && (attr as MoveEffectAttr).selfTarget && (!attr.firstHitOnly || firstHit), user, target, this.move.getMove())).then(() => {
+                  && (attr as MoveEffectAttr).selfTarget && (!attr.firstHitOnly || firstHit), user, target, move)).then(() => {
                   if (hitResult !== HitResult.NO_EFFECT) {
                     applyFilteredMoveAttrs((attr: MoveAttr) => attr instanceof MoveEffectAttr && (attr as MoveEffectAttr).trigger === MoveEffectTrigger.POST_APPLY
-                      && !(attr as MoveEffectAttr).selfTarget && (!attr.firstHitOnly || firstHit), user, target, this.move.getMove()).then(() => {
+                      && !(attr as MoveEffectAttr).selfTarget && (!attr.firstHitOnly || firstHit), user, target, move).then(() => {
```
Comment (OrangeRed) [inline src/phases.ts:2771]: Can you remove the type assertions here as well?

File: src/phases.ts [hunk 3]
```diff
@@ -2804,7 +2805,7 @@ export class MoveEffectPhase extends PokemonPhase {
         }
         // Trigger effect which should only apply one time after all targeted effects have already applied
         const postTarget = applyFilteredMoveAttrs((attr: MoveAttr) => attr instanceof MoveEffectAttr && (attr as MoveEffectAttr).trigger === MoveEffectTrigger.POST_TARGET,
-          user, null, this.move.getMove());
+          user, null, move);
```
Comment (OrangeRed) [inline src/phases.ts:2808]: Here too

**Review by OrangeRed — CHANGES_REQUESTED** [review]

File: src/field/pokemon.ts (hunk 1 above)
Comment (td76099) [reply src/field/pokemon.ts]: Removed it

File: src/phases.ts (hunk 2 above)
Comment (td76099) [reply src/phases.ts]: I am afraid to touch anything in `phases.ts`, especially that insane set of logic. Regardless, I have removed it.

File: src/phases.ts (hunk 3 above)
Comment (td76099) [reply src/phases.ts]: same as others

**Comment by td76099** [comment]
@OrangeRed Update those things and added some docs for `PokemonMove` class as discussed in Discord

Obviously tested to make sure everything still worked afterwards lol

**Review by OrangeRed — APPROVED** [review]
I think it looks good. Get one more set of eyes on it though

**Comment by star-krieg** [comment]
Tested this PR for a specific case that I read before in endless grinding chat on Discord.
Mentioning it here as well since it wasn't mentioned here.

Scrappy Slither Wing hits Wonder Guard Pokemon even if the Slither Wing had its ability overwritten with Entrainment - Normalize (Dark/Ghost and Normal/Ghost will usually die to Fighting moves enabled by Scrappy). Since Normalize would turn all of Slither Wing's moves to Normal, a Dark/Ghost Wonder Guard 'mon shouldn't be hit anymore. I overrode Slither Wing's moves to Close Combat/Focus Punch.

Here is a video of that interaction on the current version of the game:

https://github.com/pagefaultgames/pokerogue/assets/113486223/7f9fcb05-328c-4a62-9cce-d07978436c62

This is a video of the same interaction using this PR:

https://github.com/pagefaultgames/pokerogue/assets/113486223/d0ddb703-c2b5-4ef7-a4ab-1d866bc6c1b7

**Review by brain-frog — APPROVED** [review]
thanks for all the edits! great work

**Pull request merged**